Split openslides gui from main repositorie

andkit opened the pull request:
As discussed with/ wanted by @emanuelschuetze this a pull request with the changes to split out the gui from the main openslides repository. It's still somewhat WIP, but it's a start so the repo want look so empty ;)

emanuelschuetze commented:
I tested it on Windows7 32bit (with Python 3.4):
1. Download latest wxpython-phoenix snapshot from http://wxpython.org/Phoenix/snapshot-builds/ . (Use wheel file for the correct plattform, e.g. wxPython_Phoenix-3.0.3.dev1715+5ae6f4b-cp34-none-win32.whl)
2. Install wxpyhton-phoenix: `pip install wxPython_Phoenix-3.0.3.dev1715+5ae6f4b-cp34-none-win32.whl`
3. Create python package from openslides master (`python setup.py sdist`) and install it: `pip install openslides-2.0a1-dev.tar.gz`
4. clone this openslides-gui branch and go to root dir `openslides-gui`.
5. Run `python -m openslides_gui`
   => GUI comes up!*
- Note: You have to modify gui.py: remove `openslides.get_version()` because this function was removed in master and will be replaced later.

emanuelschuetze commented:
I tested it also successfully on Ubuntu 14.10 64bit (with Python 3.4):
1. Install all requirements to build wxpython_phoenix from source (there is no binary version for linux):
   see http://wiki.wxpython.org/How%20to%20install%20wxPython#Installing_wxPython-Phoenix_using_pip
   apt-get install dpkg-dev build-essential python3.4-dev libjpeg-dev libtiff-dev libgtk2.0-dev libsdl1.2-dev libgstreamer-plugins-base0.10-dev libwebkit-dev
2. Create python3.4 virtualenv: `virtualenv -p python3.4 .venv` and activate `. .venv/bin/activate`.
3. (Build and) Install wxpython-phoenix from source with pip (latest version gets syntax error): `pip install --upgrade -v --pre -f http://wxpython.org/Phoenix/snapshot-builds/ wxPython_Phoenix==3.0.3.dev1725` (needs >15min for me)
4. Clone OpenSlides master and follow the install instruction to install all requirements (into virtualenv).
5. Clone this openslides-gui branch
6. Create a symlink in openslides-master checkout to /PATHTO/openslides-gui/openslides_gui 
7. Run `python -m openslides_gui`

emanuelschuetze commented:
@andkit ping Can you update the last TODOs please? Thanks!

emanuelschuetze commented:
I tested again with wxPython_Phoenix-3.0.3.dev1820+49a8884-cp34-none-win32.whl under windows. Some feedback:
- Please comment out `openslides.get_version()` and mark it with a TODO (we have to add this function in rest api in the core).
- the OpenSlides server starts NOT correctly. If you click "Start server" a new browser tab opens but Openslides server does not run.
- "Stop server" freezed the GUI application! (You have to kill it.)
- Changing the port has no affect by starting new browser.
- Please rebase to one commit after you are ready with your changes.

emanuelschuetze commented:
Please note: Oskar und Norman changed the management command (e.g. start) for some months: https://github.com/OpenSlides/OpenSlides/pull/1412/files

andkit commented:
@emanuelschuetze I had a bit of time to look at your issues

Stopping issue should be fixed (with the possible exception of a race-condition when you try stopping right after starting, we probably think about a clean way to shut down the server in the future)

The other issues can't be fixed in the gui as far as I can see. 
`openslides start` no longer has options for setting host/port and hardcodes 0.0.0.0:8000
for both listening and as address opened by the browser (and using 0.0.0.0 for connect does not work, I guess that's why you thought the server wasn't running)

So, how do you want to proceed? Squash everything and merge what we have for now, or leave it open until the command line stuff in openslides:master is in a usable state again?

emanuelschuetze commented:
Thanks for your rework! I tested your updated branch:
Stopping server works now without freeze. Good.

@andkit Please fix:
- Please add a requirements.txt. At least `psutil` and `wxpython_phonix` are required to start the gui. Any more?
- Please comment out the second `openslides.get_version()` in gui.py in function `on_about_click` (line 635).
- It would be great if you can add more install and run instruction to README. (See e.g. https://github.com/OpenSlides/openslides-export/blob/master/README.rst)
- At the end, please rebase/squash everything to one commit.

@ostcar and @normanjaeckel can you make a first review of this pull request please?! Thanks.

andkit commented:
@emanuelschuetze 
- get_version() done
- requirements.txt done (I noticed that your other requirements.txt file don't include the package itself, so I have done the same here)
- wrt. the README, I'm not sure want you want me to write in there, right now you have to go through serval hoops since it requires openslides-git and wxPhoenix is still only in pre-release stage as well; if this becomes ready for end-users both should have a stable release, ideally in pypi (where we should also put the gui) and then installation simply becomes pip install openslides-gui (possibly with venv boiler plate, depending on personal preferences)
  Maybe it's best if you just add whatever you want to have in there yourself

emanuelschuetze commented:
@andkit Thanks for your work! Looks good. I'll add a README later.

@ostcar Please review now!

ostcar commented on openslides_gui/gui.py:
```diff
@@ -0,0 +1,763 @@
+#       a defined settings module
+_translations = gettext.NullTranslations()
+_ = lambda text: _translations.gettext(text)
+ngettext = lambda msg1, msg2, n: _translations.ngettext(msg1, msg2, n)
```
Proposal: Use:

``` python
_ = _translations.gettext
ngettext = _translations.ngettext
```

andkit replied:
That won't work since the _translations object will be replaced after the translation are loaded (not that we currently have translations)
I did however remove the obsolete comment and change the location from which translations would be loaded (during early devolpment it was an submodule of openslides and would have reused it's translations, but that never went live)

ostcar replied:
ok

ostcar commented on openslides_gui/gui.py:
```diff
@@ -0,0 +1,763 @@
+
+        self.last_backup = datetime.datetime.utcnow()
+
+    def on_syncdb_clicked(self, evt):
```
For OpenSlides 2.0 we use Django >=1.7 and therefore use the command migrate instead of syncdb.

ostcar commented:
ok

emanuelschuetze commented on openslides_gui/gui.py:
```diff
@@ -0,0 +1,763 @@
+
+        # TODO: get_version was removed - replace with new method
+         #version_str = _("Version {0}").format(openslides.get_version())
+        version_str = _("Version {0}").format("???")
```
Please use for version_str:

``` python
import openslides
print(openslides.__version__)
```

see https://github.com/OpenSlides/OpenSlides/issues/1492#issuecomment-99356146

emanuelschuetze commented on openslides_gui/gui.py:
```diff
@@ -0,0 +1,763 @@
+        info = wx.adv.AboutDialogInfo()
+        info.SetName("OpenSlides")
+        # TODO: get_version was removed - replace with new method
+        # info.SetVersion(openslides.get_version())
```
Please use for version_str:

``` python
import openslides
print(openslides.__version__)
```

see https://github.com/OpenSlides/OpenSlides/issues/1492#issuecomment-99356146

emanuelschuetze commented on openslides_gui/gui.py:
```diff
@@ -0,0 +1,755 @@
+        if self._host == "0.0.0.0":
+            args = ["--port", self._port]
+        else:
+            args = ["--address", self._host, "--port", self._port]
```
OpenSlides master uses djangos default commands/parameters to start server with special port/address:
`runserver [port or address:port]` (without any `--port` or `--address`)
see https://docs.djangoproject.com/en/1.8/ref/django-admin/#runserver-port-or-address-port
Can you please change and test it again? Thx.

emanuelschuetze commented:
@andkit A last important change is required (see my comment for runsever command). Hope you can fix it soon.

andkit commented:
@emanuelschuetze As discussed this evening I changed passing of the listen address to a single argument, matching the runserver command, but still kept using start instead of runserver (since the latter would not start a browser)
As noted this will not have any effect until the openslides start command is fixed

I have also squashed all the commits to a single commit, so this should be ready to merge now

emanuelschuetze commented:
My final test of this pull request:
1. Install wxPython_Phoenix *.whl file from untrusted (http) snapshot url once wx-phoenix is listed on pypi:
   `pip install -f --trusted-host http://wxpython.org/Phoenix/snapshot-builds/wxPython_Phoenix-3.0.3.dev1820+49a8884-cp34-none-win32.whl`
2. Create source package from openslides master source package (`python setup.py sdist`) and run:
   `pip install openslides-2.0a1.dev0.tar.gz`
3. Create sdist package from openslides-gui and run: `pip install openslides-gui-1.0.0.dev1.tar.gz`
4. Now you can start gui with: `openslides-gui` and run OpenSlides server.

=> ready to merge
(README will follow later.)

@andkit: Thanks again for your good work!
